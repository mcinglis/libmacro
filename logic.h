
// Copyright 2014  Malcolm Inglis <http://minglis.id.au>
//
// This file is part of Libmacro.
//
// Libmacro is free software: you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License as published by the
// Free Software Foundation, either version 3 of the License, or (at your
// option) any later version.
//
// Libmacro is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public
// License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with Libmacro. If not, see <https://gnu.org/licenses/>.


#ifndef LIBMACRO_LOGIC_H
#define LIBMACRO_LOGIC_H


#include <libpp/intersperse.h>  // PP_INTERSPERSE
#include <libpp/separators.h>   // PP_SEP_AND, PP_SEP_OR


// @public begin


// These macros are provided more for the possibility of functional
// programming with the preprocessor - not for direct use in source code.
#define AND( X, Y )         ( ( X ) && ( Y ) )
#define OR( X, Y )          ( ( X ) || ( Y ) )
#define NAND( X, Y )        ( !AND( X, Y ) )
#define NOR( X, Y )         ( !OR( X, Y ) )

// May clarify meaning in source code:
#define IMPLIES( X, Y )     ( !( X ) || ( Y ) )

// Retains the short-circuiting behavior across variable number of arguments:
#define ALL( ... )          ( PP_INTERSPERSE( PP_SEP_AND, __VA_ARGS__ ) )
#define ANY( ... )          ( PP_INTERSPERSE( PP_SEP_OR, __VA_ARGS__ ) )
#define NONE( ... )         ( !ANY( __VA_ARGS__ ) )


// @public end


#endif

