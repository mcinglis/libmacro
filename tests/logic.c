
#include <stdio.h>
#include <libtypes/types.h>

#include "../logic.h"

int main( void )
{
    printf( "AND( true, false ) = %d\n", AND( true, false ) );
    printf( "OR( 0, 9283 ) = %d\n", OR( 0, 9283 ) );
    printf( "NAND( OR( 0, false ), 76 ) = %d\n", NAND( OR( 0, false ), 76 ) );
    printf( "NOR( 0, true ) = %d\n", NOR( 982, 37 ) );
    printf( "IMPLIES( -23 < 0, 87 > -22 ) = %d\n",
            IMPLIES( -23 < 0, 87 > -22 ) );

    printf( "ALL( false, false, false, true ) = %d\n",
            ALL( false, false, false, true ) );
    printf( "ANY( false, false, false, true ) = %d\n",
            ANY( false, false, false, true ) );
    printf( "NONE( false, false, false, true ) = %d\n",
            NONE( false, false, false, true ) );

    printf( "ALL( false, false, false, false ) = %d\n",
            ALL( false, false, false, false ) );
    printf( "ANY( false, false, false, false ) = %d\n",
            ANY( false, false, false, false ) );
    printf( "NONE( false, false, false, false ) = %d\n",
            NONE( false, false, false, false ) );

    printf( "ALL( true, true, true, true ) = %d\n",
            ALL( true, true, true, true ) );
    printf( "ANY( true, true, true, true ) = %d\n",
            ANY( true, true, true, true ) );
    printf( "NONE( true, true, true, true ) = %d\n",
            NONE( true, true, true, true ) );
}

