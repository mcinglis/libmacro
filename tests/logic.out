AND( true, false ) = 0
OR( 0, 9283 ) = 1
NAND( OR( 0, false ), 76 ) = 1
NOR( 0, true ) = 0
IMPLIES( -23 < 0, 87 > -22 ) = 1
ALL( false, false, false, true ) = 0
ANY( false, false, false, true ) = 1
NONE( false, false, false, true ) = 0
ALL( false, false, false, false ) = 0
ANY( false, false, false, false ) = 0
NONE( false, false, false, false ) = 1
ALL( true, true, true, true ) = 1
ANY( true, true, true, true ) = 1
NONE( true, true, true, true ) = 0
